exercises/circular-buffer/example.jl

```diff
@@ -0,0 +1,177 @@
+module CircularBuffers
+
+using Base: @propagate_inbounds
+using Printf
+
+export CircularBuffer, capacity, isfull
+
+const OVERWRITE = false
+
+"""
+    CircularBuffer{T} <: AbstractVector{T}
+
+Circular buffer with elements of type `T`, useful as a FIFO or LIFO queue with fixed
+capacity.
+
+Optimized for inserting elements at the end¸ so `push!` performs better (~1.5x faster)
+than `pushfirst!`. Deleting elements at either end (`pop!`, and `popfirst!`) as well as
+`getindex` and `setindex!` on random elements are all efficient.
+
+---
+
+    CircularBuffer(v::Vector{T}) -> CircularBuffer{T}
+
+Construct a `CircularBuffer{T}` using `v` as the backing data without copying. Its capacity
+and length are equal to `length(v)`.
+"""
+mutable struct CircularBuffer{T} <: AbstractVector{T}
+    length::Int  # number of inserted items
+    lastindex::Int  # index in data of the last (i.e., most-recently push!ed) item
+    data::Vector{T}
+
+    CircularBuffer(v::Vector{T}) where {T} = (n = length(v); new{T}(n, n, v))
+end
+
+CircularBuffer{T}(v::Vector{T}) where {T} = CircularBuffer(v::Vector{T})
+
+"""
+    CircularBuffer{T}([undef,] capacity::Int) -> CircularBuffer{T}
+
+Construct a `CircularBuffer{T}` that stores at most `capacity` elements. If `undef` is
+provided, then the circular buffer is full; otherwise it is empty. In either case the
+underlying backing data is an uninitialized vector of length `capacity`.
+"""
+function CircularBuffer{T}(::UndefInitializer, n::Integer) where {T}
+    n >= 0 || throw(ArgumentError("Capacity must be nonnegative."))
+    return CircularBuffer(Vector{T}(undef, n))
+end
+
+function CircularBuffer{T}(n::Integer) where {T}
+    cb = CircularBuffer{T}(undef, n)
+    cb.length = 0
+    return cb
+end
+
+Base.size(cb::CircularBuffer) = (cb.length,)
+Base.empty!(cb::CircularBuffer) = (cb.length = 0; cb)
+
+"""
+    capacity(cb::CircularBuffer) -> Int
+
+Return the maximum number of elements that `cb` can store.
+"""
+capacity(cb::CircularBuffer) = length(cb.data)
+
+"""
+    isfull(cb::CircularBuffer) -> Bool
+
+Return `true` if `cb` cannot insert a new item without overwriting an old item, and `false`
+otherwise.
+"""
+isfull(cb::CircularBuffer) = length(cb) == capacity(cb)
+
+"""
+    _dataindex(cb::CircularBuffer, i::Integer) -> Int
+
+Given an index `i` of `cb`, return the corresponding index `j` of `cb.data` such that `cb[i]
+== cb.data[j]`.
+"""
+@inline function _dataindex(cb::CircularBuffer, i::Integer)
+    @boundscheck checkbounds(cb, i)
+    j = cb.lastindex - length(cb) + i
+    return j < 1 ? j + length(cb.data) : j
```

Comment:
I ended up with this here and in similar code because `mod1` was too big a performance drain (you can tell I was trying for performance with the bounds annotations). The current overhead over array access is negligible, but benchmarking showed that `mod1` was causing something like a 10x slowdown, I guess because of a division.

Reply:
Fair enough :)

Reply:
I've added a comment in the code, though, because now that you've pointed it out I see it is confusing.